### DISCUSSION TITLE
Ideas to Model MPPT Low Voltage Cutoff during low Irradiance Conditions

### DISCUSSION THREAD

**Discussion by natecostello** [pvlib/pvlib-python#1954] [Q&A] [answered]
I'm pretty new exploring pvlib and have an application for it that I would like to try but I don't really know how to get started, or if the hooks are present:

I'm coming from a mobile (RV/Boat) small array POV.  These are small arrays consisting of 2-8 12V or 24V panels driving an MPPT for battery charging/DC power.

One endless debate is whether to configure panels in a series or parallel configuration.  Generally speaking, since partial shading (due to trees, or masts, etc) would strongly favor parallel arrangements.  However there is a potentially strong argument for series configurations due to the MPPTs found in these applications.

Most MPPTs used in the mobile space are "buck" only and require a panel voltage of ≈ Vbatt + 5V.  In hot and/or cloudy and/or early or late day conditions 12V and 24V panels may not produce enough voltage to allow MPPTs to generate power.

I'd like to create and MPPT class that can operate within a modelchain that can effectively "gate" the production of power if Vmpp or Voc is below a parameter based thresh-hold due to irradiance conditions.  This would allow evaluations to compare series or parallel power generation for different locations and times of year.

Any ideas, tips, etc would be greatly appreciated.  I apologize if there are some conceptual errors on my side due to my limited experience with pvlib.

**Comment by kandersolar** [comment]
Currently, pvlib does not provide any functionality along these lines.  Enforcing MPPT limits would be a good enhancement in general, even outside of mobile applications.

Creating and inserting an MPPT class into the ModelChain isn't possible.  I think the best option for you is to run the ModelChain as usual and then post-process the results yourself.  It would look something like this:

```python
mc = ModelChain(...)
mc.run_model(weather)  # run ModelChain as usual, ignoring MPPT limits

v_oc_min = ...  # MPPT limits
mppt_impossible = mc.results.dc['v_oc'] < v_oc_min  # criteria for MPPT window
p_mp = mc.results.dc.copy()  # make a copy so that we don't modify the original variable
p_mp[mppt_impossible] = 0  # when MPPT is not possible, set power to zero
# use post-processed p_mp for the configuration comparison
```

**Reply by natecostello** [reply]
Thanks for the response, I hadn't had a chance to follow up.  I think the best way to do this is to pass a custom function to the model chain as the ac_model kwarg.  This is how I've done it for now and it appears to work well:

```python
# Custom Function
def mppt(model_chain):
    p_mp = model_chain.results.dc['p_mp']
    v_mp = model_chain.results.dc['v_mp']
    power_ac = np.empty_like(p_mp)
    # hard coding voltage limits for now
    v_start = V_MPPT_START
    v_continue = V_MPPT_CONTINUE
    v_threshold = v_start
    for i, (v, p) in enumerate(zip(v_mp, p_mp)):
        if v < v_threshold:
            power_ac[i] = 0
        else:
            power_ac[i] = p
            v_threshold = v_continue

            
    if isinstance(p_mp, pd.Series):
        power_ac = pd.Series(power_ac, index=p_mp.index)
        
    # must set value in model chain directly
    model_chain.results.ac = power_ac

# Pass to Model Chain
mcParallel = ModelChain(systemParallel, loc, aoi_model="physical", spectral_model="no_loss", ac_model=mppt)
 
```

**Reply by kandersolar** [reply, marked as answer]
Wow, I totally forgot that [User-defined models](https://pvlib-python.readthedocs.io/en/stable/user_guide/modelchain.html#user-defined-models) are supported!  Thanks for the reminder :)
